### ISSUE TITLE
#box() between lines breaks justification

### ISSUE THREAD

**Issue by hurzl** [typst/typst#4934]
### Description

Either the space before the `box` appears at end of line or space (or newline) after `box` appears at the next line. As it is, I have to remove all space before and after the `box`. But it doesn't create any output, so it shouldn't break the justification.

```
#set par(justify: true)

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
eiusmod tempor incididunt ut labore et dolore magnam aliquam quaerat
voluptatem. Ut enim aeque doleamus animo, cum corpore dolemus, #box()fieri
tamen permagna accessio potest, si aliquod aeternum et infinitum
impendere malum nobis opinemur. Quod idem licet transferre in
voluptatem, ut.
```

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by laurmaedje** [comment]
This is working as intended. If you have a box, then the space before it does not collapse. Special-casing empty boxes feels rather ad-hoc to me. Might I ask why you create an empty box in the first place?

**Comment by hurzl** [comment]
In reality I create a box with an annotation that must not effect the real text flow. How else can I do this?

**Comment by laurmaedje** [comment]
What kind of annotation? What's inside the box?

**Comment by hurzl** [comment]
Just some superscripted text. But it doesn't matter, an empty box is enough to break the justification.

**Comment by laurmaedje** [comment]
It matters because I'm trying to understand your use case and why there is an empty box in the first place. 

I'm not convinced that special casing empty boxes would be consistent. What if the box is empty, but stroked? Does the space still collapse?

**Comment by hurzl** [comment]
What is stroked? 
`#box(width:0pt, stroke:1pt, super("annotation"))`? 
Just the same. Space does not collapse, but it _should_.

Btw: Why does the space collapse when I add a ")"?: `dolemus, #box())fieri`
The ")" is in the next line, and the justification is _not_ broken Why?

**Comment by laurmaedje** [comment]
> What is stroked?

I mean something like this `#box(width: 0pt, height: 6pt, stroke: 1pt)`. It is clearly visible, so it would be weird if the space collapsed.

> Btw: Why does the space collapse when I add a ")"?: dolemus, #box())fieri
> The ")" is in the next line, and the justification is not broken Why?

The box is moved to the next line because there is no line break opportunity between it and the parenthesis.

**Comment by hurzl** [comment]
> I mean something like this `#box(width: 0pt, height: 6pt, stroke: 1pt)`. It is clearly visible, so it would be weird if the space collapsed.

No it's not visible, and if the line is broken, the space should disappear from the end of the line anyway.

**Comment by laurmaedje** [comment]
Have you actually tested this? It _is_ visible if you provide a height. But anyway, there seems to be no point in further discussion as your position seems settled.

**Comment by hurzl** [comment]
> Have you actually tested this?

Yes. Have you?

**Comment by hurzl** [comment]
A zero-width box and all space before it is kept at the end of line, despite being invisible. This breaks justification. Is that what is intended?

**Comment by laurmaedje** [comment]
> Yes. Have you?

Yes. Turns out that the rendering differs between PNG export (= web app) and PDF export... but this is already tracked in https://github.com/typst/typst/issues/2155.

> A zero-width box and all space before it is kept at the end of line, despite being invisible. This breaks justification. Is that what is intended?

The fact that it is invisible is, in my opinion, secondary. Spaces around boxes just don't collapse and carving out a special case for invisible or zero-width boxes will open up a can of worms. For instance, the next question is whether `A #box(width: 0pt) B` has one or two spaces etc. I think I get your use case (inserting content without it having any width), but I think adding special cases to how boxes work is not the solution.

**Comment by hurzl** [comment]
The only special case would be the 0pt box to be handled as white space.

Can I inhibit the line break after a `#box()` manually? -- Something like TeX's `\penalty`?

Is there a better way to insert an annotation? Or anything else (side note etc.)? All of these will have the same problem when using `#box`.

**Comment by laurmaedje** [comment]
You can inhibit the linebreak with a non-breaking space (`~`) or zero width joiner (`#sym.zwj`). 

As for better ways to insert an annotation: Depends on what exactly you're trying to achieve. You could trying removing the space before the box automatically with `h(0pt, weak: true)`, similar to how footnotes eat the space before them.

**Comment by hurzl** [comment]
I don't want to remove the space, I just want it to have no effect at all ... 
Yes, appending `#sym.zwj;` to the box seems to work (for now)
